### ISSUE TITLE
Creating a ScriptModule in another thread deadlocks

### ISSUE THREAD

**Issue by heiner** [pytorch/pytorch#17896]
## 🐛 Bug

Creating a TorchScript ScriptModule class deadlocks when done in a separate thread.

## To Reproduce

Run this script
```python
import threading

import torch

class MyModule(torch.jit.ScriptModule):
    """Instantiating this outside the main thread causes a deadlock."""

    def __init__(self):
        super(MyModule, self).__init__()

    @torch.jit.script_method
    def forward(self, action: torch.Tensor):
        return torch.ones((100, 100)) + action

def run():
    print('Creating ScriptModule ...')
    module = MyModule()
    print('done. Without deadlock.')
    module(torch.tensor(0))
    print('Ran ScriptModule')

thread = threading.Thread(target=run)

thread.start()

thread.join()
```

1. Run the above script.
2. Observe deadlock.

lldb backtrace at: https://gist.github.com/heiner/1f455205faeb8ceff151c6283442487a points at a GIL acquisition issue.

## Expected behavior

No deadlock.

Running the above script with `PYTORCH_JIT=0 python3 deadlock_bug.py` works and prints
```
Creating ScriptModule ...
done. Without deadlock.
Ran ScriptModule
```
However, with `PYTORCH_JIT=1` it never reaches the "done" part and needs to be SIGSTOP'd.

I don't quite understand the backtrace, or why this would happen given that there's only two threads, one of which wants to join the other. Creating the module in the main thread works, but that doesn't work as a workaround in the situation I'd like to use TorchScript in. Incidentally, if instead of creating a `torch.jit.ScriptModule` one just calls a `@torch.jit.script`-annotated function from the thread, everything works.

## Environment

Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.3
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.2
[conda] Could not collect

The same issue happens on Linux:

PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 410.79
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.1
[pip] torch==1.0.1.post2
[pip] torchvision==0.2.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl_fft                   1.0.10           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch                   1.0.1           py3.7_cuda9.0.176_cudnn7.4.2_2    pytorch
[conda] torchvision               0.2.1                      py_2    pytorch

**Comment by driazati** [comment]
This runs fine for me on master, there was a GIL re-acquisition issue in pybind11 that they've since fixed and we updated to in #17143

**Comment by suo** [comment]
Closing as this issue has been fixed.
